File: src/main/kotlin/com/xs0/gqlktx/ann/GQLArg.java
package com.xs0.gqlktx.ann;

import java.lang.annotation.*;

@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.PARAMETER)
public @interface GQLArg {
    /**
     * The name of the argument as exposed in GraphQL schema.
     */
    String value() default "";

    /**
     * The default value for this argument (in GraphQL syntax). For example,
     * "null", "true", "123", "[ RED, BLUE ]", etc. Exclusive with required=true.
     */
    String defaultsTo() default "";

    /**
     * Whether a value is required to be provided in the query. Exclusive with
     * defaultsTo.
     */
    boolean required() default false;


    GQLArg DEFAULTS = new GQLArg() {
        @Override
        public Class<? extends Annotation> annotationType() {
            return GQLArg.class;
        }

        @Override
        public String value() {
            return "";
        }

        @Override
        public String defaultsTo() {
            return "";
        }

        @Override
        public boolean required() {
            return false;
        }
    };
}

File: src/main/kotlin/com/xs0/gqlktx/ann/GQLArg.kt
package com.xs0.gqlktx.ann

@Retention(AnnotationRetention.RUNTIME)
@Target(AnnotationTarget.VALUE_PARAMETER)
annotation class GQLArg(
    /**
     * The name of the argument as exposed in GraphQL schema.
     */
    val value: String = "",
    /**
     * The default value for this argument (in GraphQL syntax). For example,
     * "null", "true", "123", "[ RED, BLUE ]", etc. Exclusive with required=true.
     */
    val defaultsTo: String = "",
    /**
     * Whether a value is required to be provided in the query. Exclusive with
     * defaultsTo.
     */
    val required: Boolean = false
)
